// fumadocs-mdx@12.0.3/dist/bun/index.js
import {
  buildConfig
} from "../chunk-U4MQ44TS.js";
import {
  createMdxLoader
} from "../chunk-RMDXSZYE.js";
import "../chunk-QAUWMR5D.js";
import "../chunk-LMG6UWCL.js";
import "../chunk-IQAEAI4P.js";
import {
  findConfigFile,
  resolvedConfig
} from "../chunk-XMFLD5J6.js";
import "../chunk-VWJKRQZR.js";

// src/bun/index.ts
import { parse } from "querystring";
import { pathToFileURL } from "url";
function createMdxPlugin(options = {}) {
  const { configPath = findConfigFile() } = options;
  async function getMdxLoader() {
    const importPath = pathToFileURL(configPath).href;
    const out = buildConfig(await import(importPath));
    return createMdxLoader(resolvedConfig(out));
  }
  return {
    name: "bun-plugin-fumadocs-mdx",
    setup(build) {
      const mdxLoader = getMdxLoader();
      build.onLoad({ filter: /\.mdx(\?.+?)?$/ }, async (args) => {
        const [filePath, query] = args.path.split("?", 2);
        const content = await Bun.file(filePath).text();
        const result = await (await mdxLoader)({
          source: content,
          query: parse(query),
          filePath,
          development: false,
          compiler: {
            addDependency() {
            }
          }
        });
        return {
          contents: result.code,
          loader: "js"
        };
      });
    }
  };
}
export {
  createMdxPlugin
};
